core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -277,22 +323,39 @@ public boolean tryComplete() {
             return false;
         } catch (Exception e) {
             log.error("Error processing delayed share fetch request", e);
-            releasePartitionLocks(topicPartitionData.keySet());
-            partitionsAcquired.clear();
-            partitionsAlreadyFetched.clear();
-            return forceComplete();
+            // In case we have a remote fetch exception, we have already released locks for partitions which have potential
+            // local log read. We do not release locks for partitions which have a remote storage read because we need to
+            // complete the share fetch request in onComplete and if we release the locks early here, some other DelayedShareFetch
+            // request might get the locks for those partitions without this one getting complete.
+            if (remoteStorageFetchException.isEmpty()) {
+                releasePartitionLocks(topicPartitionData.keySet());
+                partitionsAcquired.clear();
+                localPartitionsAlreadyFetched.clear();
+                return forceComplete();
+            } else {
+                boolean completedByMe = forceComplete();
+                // If invocation of forceComplete is not successful, then that means the request is already completed
+                // hence release the acquired locks. This can occur in case of remote storage fetch if there is a thread that
```

Comment:
> then that means the request is already completed hence release the acquired locks. 

then that means the request is already completed and hence the acquired locks are already released. Ditto above.